sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -780,6 +780,8 @@ object NullPropagation extends Rule[LogicalPlan] {
           Literal.create(null, e.dataType)
         } else if (newChildren.length == 1) {
           newChildren.head
+        } else if (!newChildren.head.nullable) {
```

Comment:
can we generalize it? I think we should find the first non-nullable child, and drop all other children after it.
```
else {
  Coalesce(newChildren.take(newChildren.indexWhere(_.nullable) + 1))
}
```